### ISSUE TITLE
ODBC dictoinary error. Can not start CH

### ISSUE THREAD

**Issue by EvgenyVinogradov** [yandex/ClickHouse#3108]
Hello!

I have upgraded CH from 18.5.1 -> 18.12.13 and can not start CH due to ODBC dictionary error:

===========================
ODBC Diagnostic record #1:
===========================
SQLSTATE = 01000
Native Error Code = 0
**[unixODBC][Driver Manager]Can't open lib '/var/lib/…/instantclient_12_2/lib/libsqora.so.12.1' : file not found**

, e.what() = Connection attempt failed, e.what() = DB::Exception, Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x9438986]
1. /usr/bin/clickhouse-server(DB::receiveResponse(Poco::Net::HTTPClientSession&, Poco::Net::HTTPRequest const&, Poco::Net::HTTPResponse&)+0x191) [0x9450e81]
2. /usr/bin/clickhouse-server(DB::ReadWriteBufferFromHTTP::ReadWriteBufferFromHTTP(Poco::URI const&, std::__cxx11::basic_string<char, std::char_traits<char>,
 std::allocator<char> > const&, std::function<void (std::ostream&)>, DB::ConnectionTimeouts const&, Poco::Net::HTTPBasicCredentials const&, unsigned long)+0x
296) [0x9469016]
3. /usr/bin/clickhouse-server() [0x76a4c0f]
4. /usr/bin/clickhouse-server(DB::ODBCDictionarySource::loadBase(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) cons
t+0x170) [0x76a4fb0]
5. /usr/bin/clickhouse-server(DB::ODBCDictionarySource::loadAll()+0x107) [0x76a51c7]
6. /usr/bin/clickhouse-server(DB::FlatDictionary::loadData()+0x4e) [0x6f4e0ae]
7. /usr/bin/clickhouse-server(DB::FlatDictionary::FlatDictionary(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::
DictionaryStructure const&, std::unique_ptr<DB::IDictionarySource, std::default_delete<DB::IDictionarySource> >, DB::ExternalLoadableLifetime, bool, std::sha
red_ptr<DB::Block>)+0x5da) [0x6f4e87a]
8. /usr/bin/clickhouse-server(DB::DictionaryFactory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Poco::Uti
l::AbstractConfiguration const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::Context&) const+0x1310) [0x75628
20]
9. /usr/bin/clickhouse-server(DB::ExternalDictionaries::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Poco::
Util::AbstractConfiguration const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x43) [0x707df63]
10. /usr/bin/clickhouse-server(DB::ExternalLoader::reloadFromConfigFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const
&, bool, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x12fb) [0x708163b]
11. /usr/bin/clickhouse-server(DB::ExternalLoader::reloadFromConfigFiles(bool, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<
char> > const&)+0xbb) [0x7083bfb]
12. /usr/bin/clickhouse-server(DB::ExternalLoader::reloadAndUpdate(bool)+0x61) [0x70842d1]
13. /usr/bin/clickhouse-server(DB::ExternalLoader::init(bool)+0x65) [0x70859c5]
14. /usr/bin/clickhouse-server(DB::ExternalDictionaries::ExternalDictionaries(std::unique_ptr<DB::IExternalLoaderConfigRepository, std::default_delete<DB::IE
xternalLoaderConfigRepository> >, DB::Context&, bool)+0x1d0) [0x707e660]
15. /usr/bin/clickhouse-server(DB::Context::getExternalDictionariesImpl(bool) const+0xf0) [0x70331c0]
16. /usr/bin/clickhouse-server() [0x721b20a]
17. /usr/bin/clickhouse-server(DB::StorageFactory::get(DB::ASTCreateQuery&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > c
onst&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::
allocator<char> > const&, DB::Context&, DB::Context&, DB::ColumnsDescription const&, bool, bool) const+0x1fe) [0x722401e]
18. /usr/bin/clickhouse-server(DB::createTableFromDefinition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__c
xx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>
 > const&, DB::Context&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x10a) [0x69f118a]
19. /usr/bin/clickhouse-server() [0x74d68b6]
20. /usr/bin/clickhouse-server() [0x74d6f84]
21. /usr/bin/clickhouse-server(ThreadPool::worker()+0x19e) [0x9639a3e]
22. /usr/bin/clickhouse-server() [0x9e0c4ef]
23. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7f52f126f494]
24. /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f52f08a1acf]

**Comment by alesapin** [comment]
Stupid question: Does file `/var/lib/…/instantclient_12_2/lib/libsqora.so.12.1` exists on file system? 
Can you provide your version of unixodbc, `/etc/odbc.ini` and `/etc/odbcinst.ini`?

**Comment by EvgenyVinogradov** [comment]
Yes, sure library file exists and it works on 18.5.1 without any ODBC reconfiguration.

root@clickhouse-1:logs# apt-cache madison unixodbc
  unixodbc |    2.3.4-1 | http://deb.debian.org/debian stretch/main amd64 Packages

cat /etc/odbc.ini

[customers]
Driver=/usr/local/lib/libmaodbc.so
Server=mysql-slave-reports
Database=customers
User=...
Password=...
Port=3306

[operators]
Driver=/usr/local/lib/libmaodbc.so
Server=mysql-slave-reports
Database=operators
User=...
Password=...
Port=3306

[aggregated]
Driver=/usr/local/lib/libmaodbc.so
Server=mysql-slave-reports
Database=aggregated
User=…
Password=...
Port=3306

[rator_ro]
Driver=/var/lib/clickhouse/utils/instantclient_12_2/lib/libsqora.so.12.1
ServerName=RATOR_PROD_RO
UserID=...
Password=…


cat /etc/odbcinst.ini

[ODBC]
Trace = yes
TraceFile = /var/lib/clickhouse/utils/odbc.trc

[MySQL]
Description=ODBC for MariaDB
Driver=/usr/local/lib/libmaodbc.so
Setup=
Driver64=/usr/local/lib/libmaodbc.so
Setup64=
FileUsage=1
UsageCount=13

[OracleODBC12c]
Description = Oracle ODBC driver for Oracle 12c
Driver=/var/lib/clickhouse/utils/instantclient_12_2/lib/libsqora.so.12.1
FileUsage = 1
Driver Logging = 0

**Comment by alesapin** [comment]
In ClickHouse 18.12.13 we have changed interaction with ODBC managers. Previously .so libraries were linked with clickhouse-server binary, now they link with clickhouse-odbc-bridge binary. `file not found` exception doesn't mean, that the file doesn't exist on fs. It may mean that shared object cannot be loaded. 

Please, provide `ldd /var/lib/clickhouse/utils/instantclient_12_2/lib/libsqora.so.12.1`.

Also in clickhouse-server.log with trace log-level you can see line like this before your exception :
`<Trace> ReadWriteBufferFromHTTP: Sending request to http://localhost:9018/?connection_string=DSN%3DClickHouse%3BDATABASE%3Dstat&columns=columns%20format%20version%3A%201%0A3%20columns%3A%0A%60PageID%60%20String%0A%60ImpID%60%20String%0A%60AdType%
60%20String%0A&max_block_size=65536`

To debug dynamic linkage, you can run:
`$ LD_DEBUG=all clickhouse-odbc-bridge --http-port 9018`
`$ curl -d 'query=select 1' -X POST 'your_url_in_log'`
In output you will see missing library.
